server/api/signup.js

```diff
@@ -1,51 +1,102 @@
+    return;
+  }
+
+  // test username and password; should throw an error if there's an issue
+  try {
+    checkUsernameRequirements(req.body.username);
+    checkPasswordRequirements(req.body.password);
+  } catch (error) {
+    res.status(401).send(error.message);
+    return;
+  }
 
-    var salt = Crypto.randomBytes(16);      // salt is required for hashing; it makes 
-                                            // otherwise identical hashes different
-    Crypto.pbkdf2(req.body.password, salt, 310000, 32, 'sha256', function(err, hashedPassword) {
+  const salt = Crypto.randomBytes(16); // salt is required for hashing; it makes
+  // otherwise identical hashes different
+  Crypto.pbkdf2(
+    req.body.password,
+    salt,
+    310000,
+    32,
+    "sha256",
+    async function (err, hashedPassword) {
       if (err) {
-        res.status(500).send(err); 
-        return next(err); 
+        res.status(500).send(err);
+        return;
       }
+
       // insert new user into database
-      db.run('INSERT INTO User (username, hashed_password, salt) VALUES (?, ?, ?)', [
-        req.body.username,
-        hashedPassword,
-        salt
-      ], function(err) {
-        if (err) {
-          // TODO: Send more descriptive errors
-          res.status(500).send(err) 
-          return next(err); 
-        }
-        var user = {
-          id: this.lastID,
-          username: req.body.username
-        };
-        // TODO: After correctly signing up, log the user in
-
-        res.status(200).send(`Successfully signed user ${user.username} up!`)
-      });
-    });
-  };
-  
-  module.exports = signup;
\ No newline at end of file
+      try {
+        const newUser = await insertUser(
+          db,
+          req.body.username,
+          hashedPassword,
+          salt,
+          (user) => {
+            res
+              .status(200)
+              .send(`Successfully signed user ${req.body.username} up!`);
+          }
+        );
+        // TODO (integration): After correctly signing up, log the user in
+      } catch (err) {
+        res
+          .status(500)
+          .send(`Error: Failed to insert new user!\nSpecific error: ${err}`);
+        return;
+      }
+    }
+  );
+}
+
+/**
+ * Inserts a new User into the database `db`. Calls a callback on the new user object afterwards.
+ * @param {Database} db - The database to insert the User into.
+ * @param {String} username - The username of the new User.
+ * @param {Buffer} hashedPassword - The hashed password of the new User. This (hopefully) should be generated by `Crypto.pbkdf2()`.
+ * @param {String} salt - The salt used in hashing the password.
+ * @param {Callback} callback - A callback which handles what happens after inserting the user.
+ *                   Should take in one argument: the `user` object generated from `insertUser`.
+ * @returns Nothing.
+ */
+async function insertUser(db, username, hashedPassword, salt, callback) {
+  await db.run(
```

Comment:
Before inserting the username + password, you should first check if there is an existing row with the same username. If there is, send an error and return from the method. If there isn't, then continue on and add the new user.

As is, the server (and app) crashes if you attempt to add a user with an existing username.
As an example, send a POST request to `/api/signup` with
```
{
    "username": "Test1234",
    "password": "Test1234"
}
```
twice, and the server will crash

Reply:
Resolved in [079b59c](https://github.com/PistolRcks/web-puzzle-project/pull/24/commits/079b59c306c8712d40520e33c6b90f92a121b814). Callback within `insertUser` now returns an error object if an error is thrown by `db.run()`.